src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2ConnectionTests.cs

```diff
@@ -26,6 +26,208 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Core.Tests
 {
     public class Http2ConnectionTests : Http2TestBase
     {
+        [Fact]
+        public async Task FlowControl_ParallelStreams_FirstInFirstOutOrder()
+        {
+            var tcs = new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+            await InitializeConnectionAsync(async c =>
+            {
+                // Send headers
+                await c.Response.Body.FlushAsync();
+
+                // Wait before sending data
+                await tcs.Task;
+
+                // Send large data (3 larger than window size)
+                await c.Response.Body.WriteAsync(new byte[65538]);
+            });
+
+            await StartStreamAsync(1, _browserRequestHeaders, endStream: true);
+            // Ensure the stream window size is large enough
+            await SendWindowUpdateAsync(streamId: 1, 65538);
+
+            await ExpectAsync(Http2FrameType.HEADERS,
+                withLength: 33,
+                withFlags: (byte)Http2HeadersFrameFlags.END_HEADERS,
+                withStreamId: 1);
+
+            tcs.SetResult(null);
+
+            await ExpectAsync(Http2FrameType.DATA,
+                withLength: 16384,
+                withFlags: (byte)Http2DataFrameFlags.NONE,
+                withStreamId: 1);
+            await ExpectAsync(Http2FrameType.DATA,
+                withLength: 16384,
+                withFlags: (byte)Http2DataFrameFlags.NONE,
+                withStreamId: 1);
+            await ExpectAsync(Http2FrameType.DATA,
+                withLength: 16384,
+                withFlags: (byte)Http2DataFrameFlags.NONE,
+                withStreamId: 1);
+            await ExpectAsync(Http2FrameType.DATA,
+                withLength: 16383,
+                withFlags: (byte)Http2DataFrameFlags.NONE,
+                withStreamId: 1);
+
+            // 3 byte is remaining on stream 1
+
+            tcs = new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+            await StartStreamAsync(3, _browserRequestHeaders, endStream: true);
+            // Ensure the stream window size is large enough
+            await SendWindowUpdateAsync(streamId: 3, 65538);
+
+            await ExpectAsync(Http2FrameType.HEADERS,
+                withLength: 33,
+                withFlags: (byte)Http2HeadersFrameFlags.END_HEADERS,
+                withStreamId: 3);
+
+            tcs.SetResult(null);
+
+            // Pause to ensure that data from the second stream is queued
+            // and waiting for window updates
+            while (_connection._outputFlowControl.ActiveAwaitableCount < 2)
```

Comment:
Another option is to set a TCS in the app func after you call WriteAsync but before you await it.

Reply:
I like that much more.